operator/pkg/apis/infrastructure/v1alpha1/controlplane.go

```diff
@@ -41,47 +42,39 @@ type ControlPlaneList struct {
 // master and etcd are configured to run. By default, KIT uses all the default
 // values and ControlPlaneSpec can be empty.
 type ControlPlaneSpec struct {
-	Infrastructure    `json:",inline"`
 	KubernetesVersion string     `json:"kubernetesVersion,omitempty"`
 	Master            MasterSpec `json:"master,omitempty"`
 	Etcd              ETCDSpec   `json:"etcd,omitempty"`
 }
 
-type Infrastructure struct {
-	VPCCidr        string   `json:"vpccidr,omitempty"`
-	Zones          []string `json:"zones,omitempty"`
-	PrivateSubnets []string `json:"privateSubnets,omitempty"`
-}
-
 // MasterSpec provides a way for the user to configure master instances and
 // custom flags for components running on master nodes like apiserver, KCM and
 // scheduler.
 type MasterSpec struct {
 	Instances         `json:",inline"`
-	Scheduler         *Config `json:"scheduler,omitempty"`
-	ControllerManager *Config `json:"controllerManager,omitempty"`
-	APIServer         *Config `json:"apiServer,omitempty"`
+	Scheduler         *Deployment `json:"scheduler,omitempty"`
+	ControllerManager *Deployment `json:"controllerManager,omitempty"`
+	APIServer         *Deployment `json:"apiServer,omitempty"`
 }
 
 // ETCDSpec provides a way to configure the etcd nodes and args which are passed to the etcd process.
 type ETCDSpec struct {
-	Instances `json:",inline"`
-	*Config   `json:",inline"`
+	Instances   `json:",inline"`
+	*Deployment `json:",inline"`
 }
 
-// Config provides a generic way to pass in args and images to master and etcd
+// Deployment provides a generic way to pass in args and images to master and etcd
 // components. If a user wants to change the QPS they need to provide the
 // following flag with the desired value -`kube-api-qps:100` in the args.
-type Config struct {
-	Args  map[string]string `json:"args,omitempty"`
-	Image string            `json:"image,omitempty"`
+type Deployment struct {
```

Comment:
Can't we use `type Deployment` directly from k8s lib ? what's additional significance we get from this ?

Reply:
Huge +1. Let's use `corev1.Deployment`

Reply:
Minimally, this buys us the ability to control the RolloutStrategy.

Reply:
The thing with using `v1.Deployment` object is it massively changes the user experience. 
This is how a user will create a default cluster-
```
apiVersion: kit.k8s.amazonaws.com/v1alpha1
kind: ControlPlane
metadata:
  name: example # Desired Cluster Name
spec: {}
```

Now if a user wants to change image name for api server and scheduler, they will need to do something like-
```
apiVersion: kit.k8s.amazonaws.com/v1alpha1
kind: ControlPlane
metadata:
  name: example # Desired Cluster Name
spec: 
  items:
  - master:
     scheduler:
       spec:
         template:
            spec:
              containers:
                 image: scheduler-image-name
     apiServer:
       spec:
         template:
            spec:
              containers:
                 image: apiserver-image-name
```
IMO, as a user I need to check the syntax for v1.Deployment everytime I need to make a small edit, at this point I might just go and edit the YAML spec in management cluster.

This field is implemented as a Kubernetes deployment object in the operator so we will be getting the rolling updates. Is there any functionality from the `v1.Deployment` spec that we need to expose to the user that we can't provide from `v1.PodSpec`?

Reply:
Speaking of rolling updates( taking a step back here), I was wondering do we actually need to update the spec of a test cluster that's already created ? I feel like that's more of an overkill to me than creating a new test cluster with the spec we wanted ? In my mind, I don't see  great benefits of updating a test cluster. Could you let me know what use-case I'm missing here  for megaxl or any test cases that we have today ?

Reply:
Use Case: test upgrading a cluster under load.

Reply:
@prateekgogia, I see your argument. We could potentially go further and do a ContainerSpec instead of a podspec, but I think we may lose useful bits like node selectors. I'm supportive of a more customized DeploymentSpec (e.g. PodSpec and Replicas)

Reply:
Cool, that could be controlled through `Replicas` count on Deployment itself (which makes sense). But is there a benefit in updating and monitoring `PodSpec` per se which is going to be crux of the controller logic IIUC ?